module-container.md

```diff
@@ -0,0 +1,312 @@
+# Module container
+
+Developing container based applications is still a fairly new topic. Even more general ideas like using containers as a module to architect applictions. Still this document tries to gather some best practices and suggests some new ideas to the community. Most of these ideas should be container runtime agnostic but still practical relevant with concrete examples. 
+
+Many of these ideas are formalised in the [Open Containers Spec](https://github.com/opencontainers/specs) but we want to give guidance for tools which are used today. We try to link to Docker, rkt and OCI examples. 
+
+> Note: This document is highly Work-In-Progress. Please get involved. Comment discuss and add your own ideas.  
+
+## Definition
+
+The term "module container" builds upon "application container" coined by Docker. An application container focuses on running a single process in contrast to multiple processes per container. If the application consists of multiple processes they are spread out to different containers. A module container refines application container with the focus on being a good building block. In addition it suggests an even smaller granularity.
+
+## Related work
+
+In this field there is much prior and contemporary art:
+
+* [The 12 Factor App](http://12factor.net/) a widely cited site describing how to write good PaaS applications. Initiated by Heroku. Many of these principles apply for containers as well.
+
+* Cloud native: In the past year the term "cloud native applications" has gained popularity. A good introduction can be found in the free ebook [Migrating to Cloud Native Application Architectures](http://pivotal.io/platform/migrating-to-cloud-native-application-architectures-ebook)
+
+The folks from [Deis](http://deis.io/) have coined the term [cluster aware image](https://twitter.com/gabrtv/status/676218070357032961).
+
+## Properties of a module container
+
+There is no canonical definition of a "module container". Instead we gather a set of guiding properties and associated best practices:
+
+1. [Linux process](#1-linux-process)
+2. [Container API](#2-container-api)
+3. [Explicit interfaces](#3-explicit-interfaces)
+4. [Disposable](#4-disposable)
+5. [Immutable](#5-immutable)
+6. [Self-Contained](#6-self-contained)
+7. [Small](#small)
+
+### 1. Linux process
+
+Before we come up with to many new ideas we should acknowledge the fact that a container is foremost a Linux process. Therefore we should apply common standards and best practices for writing Unix tools which happen to be containers. These are _React to signals_, _Use standard streams_, _Handle arguments_:
+
+#### React to signals
```

Comment:
You probably at least mention the whole pid1 problematic (or link to where is discussed). If the root process inside a (docker) container spawns child processes and they die, the parent or root gets a signal to clean that up. If this process doesn't handle that signal, the process is kept in a zombie status.